File: src/common/net/auoeke/wheel/extension/dependency/Dependency.java
package net.auoeke.wheel.extension.dependency;

import java.util.Objects;
import net.auoeke.wheel.extension.WheelExtension;

public class Dependency {
    public final String key;
    public final String artifact;
    public final String repository;

    public Dependency(String key, String artifact, String repository) {
        this.key = key;
        this.artifact = artifact;
        this.repository = repository;
    }

    @Override
    public String toString() {
        return "%s (%s)".formatted(this.key, this.artifact);
    }

    public String resolveRepository() {
        return WheelExtension.repository(this.repository);
    }

    @Override
    public boolean equals(Object object) {
        if (object == this) {
            return true;
        }

        if (object instanceof Dependency) {
            Dependency that = (Dependency) object;

            return Objects.equals(this.key, that.key)
                && Objects.equals(this.artifact, that.artifact)
                && Objects.equals(this.repository, that.repository);
        }

        return false;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.key, this.artifact, this.repository);
    }
}

File: src/net/auoeke/wheel/extension/dependency/Dependency.kt
package net.auoeke.wheel.extension.dependency

import net.auoeke.wheel.extension.WheelExtension
import java.util.*

class Dependency(val key: String, val artifact: String, val repository: String) {
    override fun equals(other: Any?): Boolean {
        return other === this || other is Dependency
            && this.key == other.key
            && this.artifact == other.artifact
            && this.repository == other.repository
    }

    override fun hashCode(): Int = Objects.hash(this.key, this.artifact, this.repository)
    override fun toString(): String = "${this.key} (${this.artifact})"

    fun resolveRepository(): String = WheelExtension.repository(this.repository)!!
}
